### ISSUE TITLE
gpus not accessible on gpu nodes

### ISSUE THREAD

**Issue by mgoldey** [FluidNumerics/slurm-gcp#8]
I've reported this to schedmd at https://bugs.schedmd.com/show_bug.cgi?id=7901, but I'm posting it here as well to alert to the problem and to also solicit any advice for other approaches.

**Comment by schoonovernumerics** [comment]
This issue is a result of recent updates to CentOS from 7.6 to 7.7. Google has not updated the base CentOS image yet. The inconsistency between the CentOS kernel version  (957) and the kernel-devel and kernel-headers (1062) leads to a failed install of Nvidia drivers.
We will push up a patch to this repository as soon as we can.

**Comment by mgoldey** [comment]
Thank you very much for your prompt and helpful response. This is *very* good for the public image of FluidNumerics.

**Comment by schoonovernumerics** [comment]
We accept pull requests on this repository if you are able to get to it before we can.

**Comment by mgoldey** [comment]
Very much understood. I spent a few days trying to diagnose this issue having assumed the issue lay with my local modifications before I tested the upstream repositories sufficiently. The install scripts from Nvidia seem to silently fail until you try to manually install the driver by itself (before installing cuda). I think that's the way to get some reliability here, but I don't actually know how to work around this at the moment. I'll update if I get time free to look again.

Essentially if this can run with exit code 0, it should be actually install the drivers correctly. This script even has a way to specify the kernel source directories, but I'm really not sure if that fixes the issue or if the incompatibility is worse than just a file path issue.
```
wget http://fr.download.nvidia.com/tesla/410.79/NVIDIA-Linux-x86_64-410.79.run
sh NVIDIA-Linux-x86_64-410.79.run --silent
```
The 418.67 driver install script should behave the same way.

**Comment by mgoldey** [comment]
I've not had the time to look at this yet. Any progress from your side by chance?

**Comment by schoonovernumerics** [comment]
The issue is due to the CentOS 7 image on Google having an incompatible kernel with the kernel-devel and kernel-headers available from rpm. Once the image is updated to 7.7 this will be resolved.

**Comment by mgoldey** [comment]
I believe I've been able to work around this issue by preparing a custom centos 7.7 image and using that in the `slurm.jinja` file.

Steps to reproduce: 
1) Launch centos 7 VM
2) Update kernel (`yum check-update` then `yum update`) and reboot - unsure how to automate in startup-script with reboot, which is the main blocker from upstreaming this process.
3) Install nvidia driver and cuda  - driver first while manually specifying the kernel source path
```
  yum -y groupinstall "GNOME Desktop" "Development Tools"
  # Install Extra Packages for Enterprise Linux (EPEL) and other  dependencies
  yum install -y yum-utils epel-release kernel-devel kernel-headers which gcc dkms wget
  yum-config-manager --add-repo http://developer.download.nvidia.com/compute/cuda/repos/rhel7/x86_64/cuda-rhel7.repo
  echo "Installing kernel headers and development packages."
  [ ! -e NVIDIA-Linux-x86_64-410.79.run ] && wget http://fr.download.nvidia.com/tesla/410.79/NVIDIA-Linux-x86_64-410.79.run
  sh NVIDIA-Linux-x86_64-410.79.run --silent --kernel-source-path="/lib/modules/$(uname -r)/source"
  yum install -y cuda
  grub2-mkconfig -o /boot/efi/EFI/centos/grub.cfg
  mv /boot/initramfs-$(uname -r).img /boot/initramfs-$(uname -r)-nouveau.img
  dracut /boot/initramfs-$(uname -r).img $(uname -r)
  grub2-mkconfig -o /boot/grub2/grub.cfg
  echo blacklist nouveau >> /etc/modprobe.d/blacklist.conf
```
4) Create an image in gcloud GUI from the VM's disk
5) Use that image in slurm.jinja as sourceImage

Steps 4 and 5 could be ignored if we could reboot in the startup-script in the middle of the setup process.

**Comment by mgoldey** [comment]
Another issue was that the default GID for slurm was already used on the gcloud VM when using Centos-7.7. Setting it to a much higher number (like `1337`) enabled launch of the cluster.
